pkg/addonmanager/controllers/agentdeploy/controller.go

```diff
@@ -131,52 +131,86 @@ func (c *addonDeployController) sync(ctx context.Context, syncCtx factory.SyncCo
 	}
 
 	managedClusterAddon, err := c.managedClusterAddonLister.ManagedClusterAddOns(clusterName).Get(addonName)
-	if errors.IsNotFound(err) {
-		c.cache.removeCache(addonName, clusterName)
-		return nil
-	}
 	if err != nil {
+		if errors.IsNotFound(err) {
+			c.cache.removeCache(addonName, clusterName)
+			return nil
+		}
 		return err
 	}
-
-	if !managedClusterAddon.DeletionTimestamp.IsZero() {
-		c.cache.removeCache(addonName, clusterName)
-		return nil
-	}
-
+	managedClusterAddonCopy := managedClusterAddon.DeepCopy()
 	owner := metav1.NewControllerRef(managedClusterAddon, addonapiv1alpha1.GroupVersion.WithKind("ManagedClusterAddOn"))
 
-	managedClusterAddonCopy := managedClusterAddon.DeepCopy()
 	objects, err := agentAddon.Manifests(managedCluster, managedClusterAddon)
 	if err != nil {
 		return err
 	}
 	if len(objects) == 0 {
 		return nil
 	}
-
-	work, err := buildManifestWorkFromObject(clusterName, addonName, objects)
+	deployWork, hookWork, err := buildManifestWorkFromObject(clusterName, addonName, objects)
 	if err != nil {
 		return err
 	}
-	work.OwnerReferences = []metav1.OwnerReference{*owner}
+
+	if hookWork != nil {
+		if !hasFinalizer(managedClusterAddonCopy.GetFinalizers(), constants.PreDeleteHookFinalizer) {
+			managedClusterAddonCopy.Finalizers = append(managedClusterAddonCopy.Finalizers, constants.PreDeleteHookFinalizer)
+			_, err = c.addonClient.AddonV1alpha1().ManagedClusterAddOns(managedClusterAddonCopy.Namespace).Update(
+				ctx, managedClusterAddonCopy, metav1.UpdateOptions{})
+			return err
+		}
+	}
+
+	if !managedClusterAddon.DeletionTimestamp.IsZero() {
+		c.cache.removeCache(addonName, clusterName)
+		if hookWork != nil {
+			// run hookWork
+			hookWork.OwnerReferences = []metav1.OwnerReference{*owner}
+			hookWork, err = c.runHookWork(ctx, hookWork)
+			if err != nil {
+				meta.SetStatusCondition(&managedClusterAddonCopy.Status.Conditions, metav1.Condition{
+					Type:    "ManifestApplied",
+					Status:  metav1.ConditionFalse,
+					Reason:  "ManifestWorkApplyFailed",
+					Message: fmt.Sprintf("failed to apply manifestwork %v: %v", hookWork.Name, err),
+				})
+				c.addonClient.AddonV1alpha1().ManagedClusterAddOns(managedClusterAddonCopy.Namespace).UpdateStatus(
+					ctx, managedClusterAddonCopy, metav1.UpdateOptions{})
+				return err
+			}
+
+			if hookWorkIsFinished(hookWork) {
+				finalizer := removeFinalizer(managedClusterAddonCopy.Finalizers, constants.PreDeleteHookFinalizer)
+				managedClusterAddonCopy.SetFinalizers(finalizer)
+				_, err = c.addonClient.AddonV1alpha1().ManagedClusterAddOns(managedClusterAddonCopy.Namespace).Update(
+					ctx, managedClusterAddonCopy, metav1.UpdateOptions{})
+				return err
+			}
+			return fmt.Errorf("reconcile since the hook manifetsWorks is not finished")
```

Comment:
I do not think we should return error here. We should monitor hookWork, and reconcile when hookwork is updated.

Reply:
done.